### DISCUSSION TITLE
API Engine situation et serialisation

### DISCUSSION THREAD

**Discussion by lajarre** [publicodes/publicodes#406] [General]
Suite au travail sur `useSearchParamsSimulationSharing.ts`, j'avais besoin de comprendre comment fonctionne l'interface entre l'engine et le code React de mon-entreprise.  
Je pense aussi au cas de figure du développeur qui veut réutiliser le package publicodes et l'Engine en particulier.

Le principal point qui n'était pas clair était: à quoi doivent ressembler les valeurs de l'objet Situation, en particulier de l'argument à `Engine.setSituation`.

On voit dans le state Redux que les valeurs pour des `numbers` sont du type `{ value: 5, unit: '€/mois' }` alors que dans l'output de `serializeEvaluation` qu'on a construit sous l'impulsion de @johangirod on a des valeurs du genre `'5 €/mois'`.

Ma principale remarque: comment donner le bon argument à setSituation?  
Dans la codebase, ça ne me semble pas clair, pas centralisé. Aussi, il me semble qu'il est assez difficle pour l'Engine d'évoluer sur ce point sans faire planter le code mon-entreprise de manière trop subtile.  

1. Je pars du principe qu'il n'y a pas de raison que la serialisation soit différente entre le `useSearchParamsSimulationSharing` et les components d'input comme `Input.tsx`. A mon avis on devrait pouvoir arranger ça.
2. Dans cet élan, pouvoir ensuite typer proprement la Situation.
3. Il est cohérent que les components d'input comme `Input.tsx` sachent qu'ils doivent transformer la valeur qu'ils envoient dans leur `onChange`, car après tout chaque component correspond à un type de valeur. Ceci dit c'est potentiellement mieux séparable, car c'est peut-être dommage d'avoir un component UI qui se pose des questions sur comment gérer le sate de l'Engine. A creuser.
4. En tant que développeur qui utilise publicodes (ici, mon-entreprise), il faut pouvoir vérifier son interface avec l'Engine et s'assurer qu'on envoie des valeurs bien formées. Ca passe par un typage correct et/ou des tests d'interface.

**Comment by johangirod** [comment]
Le typage se setSituation est pour le coup assez explicite dans [core/index.ts](https://github.com/betagouv/mon-entreprise/blob/master/publicodes/core/source/index.ts#L109). 

C'est un objet clé valeur, avec comme clé un nom de règle et comme valeur, une expression publicode bien formée, ou un noeud de l'AST parsé.

Ainsi les valeurs de type nombre avec unité peuvent s'écrire de deux façon différente en publicodes. Soit une chaîne de charactère (`5 €/mois`), soit un objet avec qui décrit l'application du mécanisme `unité` sur une valeur numérique neutre.
```json
{ "valeur": 5, "unité": "€/mois }
```

> Je pars du principe qu'il n'y a pas de raison que la serialisation soit différente entre le useSearchParamsSimulationSharing et les components d'input comme Input.tsx.

Pas forcément. `useSearchParamsSimulationSharing` a explicitement besoin d'une serialisation sous forme de chaîne de charactère, tandis que pour Input.tsx, on peut se contenter d'objets javascript serialisable (puisqu'ils seront stockés dans le state redux). Dans l'absolu, on pourrait n'utiliser que des chaînes de charactère, mais ce n'est pas forcément souhaitable.


Ce qui n'est pas clair, je pense c'est le fait qu'on puisse utiliser des expressions déjà parsée. En vrai, c'est pour deux raisons : 
1. Utiliser directement les suggestions, qui sont parsées en même temps que les règles
2. Garder une interface commune entre engine.evaluate et engine.setSituation

Il est possible que ce mélange expression parsée / non parsée soit quelque chose qui pose problème. Il est possible qu'il faille mieux délimiter ce qui a attrait au parsing et à l'évaluation. Peut-être en utilisant uniquement des expressions parsées dans Engine et avoir une classe Parser dédiée et séparée ?

**Comment by lajarre** [comment]
> Le typage se setSituation est pour le coup assez explicite dans core/index.ts.

En effet, côté publicodes c'est clair côté types. C'est côté implémentation qu'il y a `mon-entreprise/source/reducers/rootReducer.ts` =>  `export type Situation = Partial<Record<DottedName, any>>`

Ceci dit je ne suis pas encore sûr de voir pourquoi on aurait des PublicodesExpression et pas que des ASTNode

A mon avis on devrait pouvoir faire simplement un serializer / deserializer les strings de l'URL en ASTNode directement, ce qui éviterait d'avoir une abstraction supplémentaire et surtout un state qui change "sans changer" (sémantique idem, syntaxe différente).

A quels autres fins est-ce qu'on a besoin des PublicodesExpression?  
J'ai pas encore trop regarder je vais le faire.

**Comment by johangirod** [comment]
> A mon avis on devrait pouvoir faire simplement un serializer / deserializer les strings de l'URL en ASTNode directement, ce qui éviterait d'avoir une abstraction supplémentaire et surtout un state qui change "sans changer" (sémantique idem, syntaxe différente).

L'inconvénient, c'est qu'on ajoute une dépendance à des structures internes du moteur. L'autre solution c'est de passer par un phase de parsing explicite (via un appel de fonction ou autre). L'avantage c'est qu'on réduit la surface d'API de setSituation / evaluate. L'inconvénient, c'est qu'on rend l'utilisation du moteur plus verbeuse : 
```
const parser = new Parser(rule)
engine = new Engine(parser)
engine.evaluate(parser.parse('dépenses primeur'))
```

Personnellement, je pense que l'interface du moteur devrait être celle d'un [interpreteur](https://fr.wikipedia.org/wiki/Interpr%C3%A8te_(informatique)), sur le même principe que l'interpreteur nodeJS ou pyhton par exemple. Aujourd'hui c'est une interface hybride, mi-compileur mi-interpreteur. Je pense qu'on ne devrait pas exposer les ASTNode interne, et que tout devrait passer par des expressions publicodes. 

Mais c'est un avis basé sur une intuition, je n'ai pas encore réflechi plus que ça aux cas d'usage.

**Comment by denismerigoux** [comment]
Salut ! Je me permet d'intervenir dans cette discussion intéressante pour expliquer ce qui se fait classiquement en écriture de compilateurs.

Un compilateur est normalement structuré en plusieurs "représentations intermédiaire".  Chaque représentation intermédiaire permet de stocker un programme complet, mais à une phase plus ou moins avancée de la compilation. 

Dans un compilateur classique, la première représentation intermédiaire est l'"AST de surface", qui est simplement le résultat du parsing du code source sous forme de texte. De ce que je comprend vous n'avez pas d'autre représentation intermédiaire après et faites tout sur cet AST, mais par exemple dans Catala on a 4 représentations intermédiaires différentes : https://catala-lang.org/ocaml_docs/catala/index.html.

Pour effectuer la compilation et passer d'une représentation intermédiaire à une autre, un compilateur dispose généralement d'un fichier maître `driver` qui va appeler les différentes transformations à enchaîner pour faire une compilation complète. La première transformation à enchaîner est donc `parser.parser` sur un code source textuel, puis on peut enchaîner sur `engine.evaluate` qui prend l'AST de surface en arguments.

Ça c'est la manière d'organiser le code en interne, après vous pouvez exposer l'API comme vous voulez. Par exemple vous pouvez avoir `engine.evaluate_source_expr` et `engine.evaluate_ast`. Concernant le fait d'exposer une "structure interne du moteur" qui j'imagine fait référence à l'AST, c'est une pratique courante en écriture de compilateurs. En effet, les différentes représentations intermédiaires du compilateur sont autant de point d'entrées de l'API du compilateur, qui vous permet d'échanger des données avec les clients dans un format adapté sans avoir à remonter à chaque fois au code source textuel qu'il faudrait re-parser à chaque fois.

Avoir accès aux représentations intermédiaires d'un compilateur permet aussi à d'autres projets de se plugguer dessus. Par exemple, on pourrait imaginer avoir un frontend Publicodes alternatif avec une syntaxe en XML au lieu de Yaml. Du côté des exemples industriels, le compilateur LLVM à l'origine fait pour C/C++ est devenu célèbre justement grâce à sa représentation intermédiaire LLVM IR qui est maintenant utilisée comme cible de compilation par des dizaines d'autres langages.

**Comment by lajarre** [comment]
> inconvénient, c'est qu'on rend l'utilisation du moteur plus verbeuse :

Pour ça je pense qu'une micro-librairie qui fait un wrapper autour de parser + engine solutionne le problème pour les clients de publicodes.

> Aujourd'hui c'est une interface hybride, mi-compileur mi-interpreteur. Je pense qu'on ne devrait pas exposer les ASTNode interne, et que tout devrait passer par des expressions publicodes.

Je ne me sens pas très à l'aise avec ce que tu dis sur ce point @johangirod . Justement le state Redux est déjà un AST, donc la barrière est déjà franchie.

Il me semble que si on veut éviter d'exposer l'AST (pour l'instant du moins) le plus simple serait de proposer une API de state management dans l'Engine et de virer la situation du state Redux. Pour ce faire, plutôt que faire des `setSituation`, on ferait des `updateVar(DottedName, Value): Situation` ce qui me semble être l'API correspondant à l'action `UPDATE_SITUATION` dans mon-entreprise. Le seul "challenge" resterait de bien définir le type `Value` (il me semble que c'est déjà qque part).

L'autre solution serait d'exposer l'AST ou _un_ AST du moins, mais c'est peut-être pas le bon moment pour cela si on considère que l'AST n'est pas assez stable en sa version actuelle pour être exposé? Il me semble que c'est ce que tu sous-entendais dans ta remarque, @johangirod .
Et on ne veut peut-être pas se lancer maintenant dans la construction d'un autre AST qui soit plus stable et plus adapté pour l'API (même si ce n'est peut–être pas si compliqué, j'ai aussi l'impression que la réponse se trouve dans le code de `setSituation` déjà existant…).

> dans un format adapté sans avoir à remonter à chaque fois au code source textuel qu'il faudrait re-parser à chaque fois.
> 
> Avoir accès aux représentations intermédiaires d'un compilateur permet aussi à d'autres projets de se plugguer dessus. Par exemple, on pourrait imaginer avoir un frontend Publicodes alternatif avec une syntaxe en XML au lieu de Yaml.

D'accod avec tout ça @denismerigoux .  
Partager un AST est sans doute bien plus utile dans certains cas d'usages, le state Redux de mon-entreprise semble en être naturellement un.  
Le point sur le fait que Yaml n'est pas fondamental me semble juste (contrairement au cas de Python dont la syntaxe par blocs d'indentation par exemple est beaucoup plus fondamentale à son adoption par les développeurs). Donc il est bon de garder en tête qu'on veut pouvoir potentiellement exposer autre chose, dans une vision plus lointaine. Aujourd'hui, on n'a pas assez de contributeurs de bases de règles pour se projeter sur ce point.

**Comment by johangirod** [comment]
> En effet, les différentes représentations intermédiaires du compilateur sont autant de point d'entrées de l'API du compilateur, qui vous permet d'échanger des données avec les clients dans un format adapté sans avoir à remonter à chaque fois au code source textuel qu'il faudrait re-parser à chaque fois.

> Avoir accès aux représentations intermédiaires d'un compilateur permet aussi à d'autres projets de se plugguer dessus. Par exemple, on pourrait imaginer avoir un frontend Publicodes alternatif avec une syntaxe en XML au lieu de Yaml. Du côté des exemples industriels, le compilateur LLVM à l'origine fait pour C/C++ est devenu célèbre justement grâce à sa représentation intermédiaire LLVM IR qui est maintenant utilisée comme cible de compilation par des dizaines d'autres langages.

> L'autre solution serait d'exposer l'AST ou un AST du moins, mais c'est peut-être pas le bon moment pour cela si on considère que l'AST n'est pas assez stable en sa version actuelle pour être exposé

Je suis convaincu; l'AST doit donc une structure publique et documentée :+1:. D'où le besoin de converger très rapidement vers une structure assez perenne pour être publiée (ce qui devrait être le cas, une fois que les issues suivantes seront reglées : betagouv/publicodes#23, betagouv/publicodes#33, betagouv/publicodes#14)

D'ailleurs, la doc catala est très claire à ce sujet, on pourrait s'en inspirer.

> Il me semble que si on veut éviter d'exposer l'AST (pour l'instant du moins) le plus simple serait de proposer une API de state management dans l'Engine et de virer la situation du state Redux. Pour ce faire, plutôt que faire des setSituation, on ferait des updateVar(DottedName, Value): Situation

Je ne suis pas contre avoir la possibilité de faire des modifications incrémentales de la situation dans le moteur, à partir du moment où ce dernier a déjà un état interne. Ceci étant dit, je pense que dans le cas de mon-entreprise, il n'est pas totalement déconnant d'avoir la situation dans le state redux (pour faire des manipulations et des traitements dessus). Quitte à utiliser directement une version parsée pour plus de cohérence.
